Support UDP/TCP port fowarding to a host without setting up a tun

cre4ture opened the pull request:
This implements #1014.
It also allows to use remote SOCKS proxy (or `ssh` server) to enable the forwarding of more than a single port at a time.
This would solve the needs for #915.

Build-Status twin PR: https://github.com/cre4ture/nebula/pull/1

How to be used can be read in the example `config.yml`:

```yml
# by using port port forwarding (port tunnels) its possible to establish connections
# from/into the nebula-network without using a tun/tap device and thus without requiring root access
# on the host. port forwarding is only supported when setting "tun.user" is set to true and thus
# a user-tun instead of a real one is used.
# IMPORTANT: For incoming tunnels, don't forget to also open the firewall for the relevant ports.
port_forwarding:
  outbound:
  # format of local and remote address: <host/ip>:<port>
  #- local_address: 127.0.0.1:3399
  #  remote_address: 192.168.100.92:4499
     # format of protocols lists (yml-list): [tcp], [udp], [tcp, udp]
  #  protocols: [tcp, udp]
  inbound:
  # format of forward_address: <host/ip>:<port>
  #- port: 5599
  #  forward_address: 127.0.0.1:5599
     # format of protocols lists (yml-list): [tcp], [udp], [tcp, udp]
  #  protocols: [tcp, udp]

```

~Till now I only did some basic manual tests with netcat~. They resulted in the expected behavior.
Further tests where done by @sybrensa and also by myself.
- [x] outgoing TCP tested intensively
- [x] ~usafe_routes~ FAILS, but potential fix for the user-tun available here: https://github.com/cre4ture/nebula/tree/feature/enable_unsafe_routes_for_user_tun
- [x] incoming TCP tested intensively
- [x] incoming UDP tested intensively (I used iperf3. Datarate was reduced, but works)
- [x] outgoing UDP tested intensively
- more performance tuning (different PR I guess)

~These points are still open:~
- [x] namings like "tunnel" vs. "forwarding", "ingoing" vs. "inbound", ...  Please help me in finding the right terminology as I'm not an expert in this topic.
- [x] mutexes needed? I honestly did't yet explicitly check for potential concurrent accesses. this is ~a TODO~ DONE.
- [x] clean shutdown - Currently it gives some errors when shutting down. this is ~also TODO~ DONE.
- [x] memory leaks - ~I think~ DONE.
- [x] performance tunings - DONE, but could be more
- [x] 2nd iteration of performance tunings - DONE, but could be more
- [x] reloadable (SIGHUP) port forwarding configuration

I did file copy tests. ~What I got was 1/2 the rate and around 4 times more CPU usage compared to the case with kernel-tun.
I tried multiple different improvements on my code, but I fear its actually more a limitation of the gvisor netstack :-/~
Any help or idea is welcome.

Update: My file copy tests achieve now around 90% of the speed at 160% of CPU load of nebula. This is a significant improvment and acceptable (at least for my useases). Thanks @akernet for the support here.

akernet committed:
Add script for speedtesting port forwarding

akernet commented:
Thanks a ton for taking this project on!

I took a stab at this before finding your changes here, it's [here](https://github.com/slackhq/nebula/compare/master...akernet:nebula:push-ynqokntmuslt) if you want to take a look.

In particular [e764eb](https://github.com/slackhq/nebula/commit/e764eb4f5cfe46b8375baa820d6fc9726a58b032) adds a script that sets up a two device loopback tunnel (without root) and runs an iperf3 speed test.

I was able to improve throughput by ~7x by adding some buffering to the `UserDevice` ([a16fdb](https://github.com/slackhq/nebula/commit/a16fdb5b0792a19e24ba504fb8a7dc8f585a0dc0)) which you might want to try too. With this change it is running at 1.6 Gbit/s with a 1300 MTU which makes it quite usable to me at least. 

I still think there are several improvements left on the table:
 - Re-use allocations in this buffering by using something like `sync.Pool`, 
 - Or even better, avoid these copies all together, although that would require some bigger changes in the rest of Nebula.
 - Depending on how the internal locking of gvisor works you could probably gain some by running several net stacks in parallel that handle their individual set of connections (distributing packets based on their source address+destination port). This would not improve single socket connections though.

cre4ture commented:
> Thanks a ton for taking this project on!
> 
> I took a stab at this before finding your changes here, it's [here](https://github.com/slackhq/nebula/compare/master...akernet:nebula:push-ynqokntmuslt) if you want to take a look.
> 
> In particular [e764eb](https://github.com/slackhq/nebula/commit/e764eb4f5cfe46b8375baa820d6fc9726a58b032) adds a script that sets up a two device loopback tunnel (without root) and runs an iperf3 speed test.
> 
> I was able to improve throughput by ~7x by adding some buffering to the `UserDevice` ([a16fdb](https://github.com/slackhq/nebula/commit/a16fdb5b0792a19e24ba504fb8a7dc8f585a0dc0)) which you might want to try too. With this change it is running at 1.6 Gbit/s with a 1300 MTU which makes it quite usable to me at least.
> 
> I still think there are several improvements left on the table:
> 
>     * Re-use allocations in this buffering by using something like `sync.Pool`,
> 
>     * Or even better, avoid these copies all together, although that would require some bigger changes in the rest of Nebula.
> 
>     * Depending on how the internal locking of gvisor works you could probably gain some by running several net stacks in parallel that handle their individual set of connections (distributing packets based on their source address+destination port). This would not improve single socket connections though.

Hey, cool. I could use some support especially regarding the performance and the automated testing. But also for other golang specifics as I'm rather unexperienced there.

I tried to do some performance profiling with the help of `pprof`.
This results where leading me to problems in gvisor itself, as I can read it from this diagram:
![grafik](https://github.com/user-attachments/assets/e4b226c6-80e0-4fa7-8de2-9fdec8eb4104)

It seems that its doing some dynamic memory allocation. But from a first glance into the relevant source-code I could not see the reason for it. Do you have an idea? I'm actually already failing to upgrade the gvisior to the latest version. something is wrong with it such that it complains about multiple packages in a directrory. But how can this be undiscovered by the maintainters of gvisor? I have the feeling that I'm doing something wrong... :-/

I will for sure try your idea with the buffered queues. 
I didn't yet look in detail into your branch. But it seems that we had some similar approach using the gvisor stack.
I would be glad, if you can do a review of my code. It seems that the maintainers of the repo where so far bussy with other stuff. ;-)

johnmaguire commented:
Hi @cre4ture - this PR looks very interesting! I wanted to let you know that the maintainers best suited to review this PR are currently working on #6 which requires some major rework of Nebula, so it will probably be a bit before we are able to dig in. That being said, thanks for the contribution!

cre4ture commented:
@akernet I added your test and afterwards also the performance improvement via cherry-picking. Hope this is fine for you :-)

Your performance improvement lead to a improvment of x6 for the test in your testscript-commit on my hardware. I will do a further test where I will use this on a real-world example (the file-copy test with two seperate machines) to confirm this improvment. I'm a bit concerned because the pprof profiling directed my in some different direct. At least this is how I interpret it right now. I will net you know.

cre4ture commented:
@akernet 

it seems to me that the improvement with the buffered pipes has no impact on the my testing scenario with the file-copy between two different machines. I'm using a "private cloud" server storage called ["garage"](https://garagehq.deuxfleurs.fr/). The two machines are my laptop and a NUC PC ("server"). These are my results:

```
1. test-case: official released binary - kernel tun:
95.5 MB/second
CPU load on server side:
88 % nebula
71 % garage

2. test-case: local build with buffered pipes - gvisor tun+stack:
52.8 MB/second
CPU load server side:
197 % nebula
34 % garage

2. test-case: local build no buffered pipes - gvisor tun+stack:
52.3 MB/second
CPU load server side:
195 % nebula
35 % garage

3. test-case: local build with the buffered pipes - but kernel tun (as in 1. test-case):
99.5 MB/second
94% nebula
72% garage
```

The last test shall demonstrate the the locally compiled binary has comparable performance as the release binary.
So what exactly this means is not yet clear to me. It seems that the buffered pipes do not have a significant impact for this test-scenarios. But I can't reason about it.

akernet commented:
> I would be glad, if you can do a review of my code. It seems that the maintainers of the repo where so far bussy with other stuff. ;-)

I'm also new to go but I'll try to find some time in the next days! :)

>I added your test and afterwards also the performance improvement via cherry-picking. Hope this is fine for you :-)

Ofc!

>I'm using a "private cloud" server storage called ["garage"](https://garagehq.deuxfleurs.fr/)

Cool, I've been looking at garage for the past weeks so nice to see that you are using it with Nebula!

>It seems that the buffered pipes do not have a significant impact for this test-scenarios. But I can't reason about it.

Yeah this is a bit strange. How is the system load overall, is it close to max? The buffering should not make Nebula more efficient when it comes to CPU, in fact it's probably gonna be more costly due to the copying. What it does is decoupling the outside and inside (gvisor) parts, allowing them to run in parallel. If the system is at limits already this is unlikely to help, however. I'll try to do some benchmarking in CPU limited scenarios too, gvisor is always gonna use some additional resources but the 66% of your pprof run seems a bit high.

cre4ture commented:
Update regarding performance:
Yesterday I could achieve a significant improvement for my filecopy test-scenario:
```
"performance 5":

89.8 MB/second
cpu-load server side:
163 % CPU nebula
63 % CPU garage
```

This brings the copy-speed to almost the reference (99MB/second) - only 10% difference remaining.
The CPU-load of nebula is only + ~60% now.

So overall we are at around 170% difference. Which is significant compared to the initial ~ 400%.

@akernet I think the main difference comes from a further improvement of your performance tuning "buffering to UserDevice". I achieved it by addtionally avoiding data-copy and dynamic allocation steps (i think).

akernet commented on port-forwarder/fwd_tcp.go:
```diff
@@ -0,0 +1,230 @@
+	if err != nil {
+		return nil, err
+	}
+	remoteTcpAddr, err := net.ResolveTCPAddr("tcp", cf.remoteConnect)
```
I would recommend that you do this resolution when needed. For IPs it doesn't really matter, but host lookups might change after starting Nebula.

Another thing that I don't think we have to cover in this PR is looking up IPs from the Nebula network instead of using DNS

cre4ture replied:
Ah, I see. Your right. I never used it, but nebula has an experimental DNS, right?

cre4ture replied:
done. thanks

akernet commented on port-forwarder/fwd_tcp.go:
```diff
@@ -0,0 +1,230 @@
+	}
+}
+
+func (pt *PortForwardingOutgoingTcp) handleClientConnection_intern(localConnection net.Conn) error {
```
Could we remove this function? Looks a bit redundant

cre4ture replied:
I like to keep it. There is a tiny but significant advantage of keeping it: The logging for the error case needs to be implemented only once instead of twice when merging the function into the caller function. I renamed the function to make this a bit more clear.

akernet commented on port-forwarder/fwd_tcp.go:
```diff
@@ -0,0 +1,230 @@
+		defer to.Close()
+		defer cancel()
+		// give communication in opposite direction time to finish as well
+		defer time.Sleep(time.Millisecond * 300)
```
I think you want to avoid sleeps like this one if possible

cre4ture replied:
I agree, especially in other languages like C++ where the default is synchronous execution. With golang I wasn't so sure whether its a real problem as the goroutines seems to compare to the async sleeps of other languages.

I honestly do not know it this is actually really needed at all. Because this is intented to catch the case where the receiving side is closed before the sending side could send all requested data. I do not have a testing possibility for this. Do you know if this is needed, or how to test it ideally?

cre4ture replied:
I removed it.

akernet commented on port-forwarder/fwd_tcp.go:
```diff
@@ -0,0 +1,230 @@
+			default:
+			}
+
+			rn, r_err := from.Read(buf)
```
I wonder if this could be replaced with `io.Copy()`

cre4ture replied:
it can. I once tried this. But it didn't improve the performance. Thats why I went on with other experiments. But now, as we solved the issue to a big extent, I will introduce it again as it simplifies the code. Thanks for pointing it out.

cre4ture replied:
I added a variant with io.Copy().
Problem: It seems to slightly decrease performance. So I'm wondering if I better should keep the original implementation.

akernet commented on port-forwarder/fwd_tcp.go:
```diff
@@ -0,0 +1,230 @@
+	return fmt.Sprintf("inbound.tcp.%d.%s", cfg.port, cfg.forwardLocalAddress)
+}
+
+type PortForwardingOutgoingTcp struct {
```
Suggestion: setup a `context.WithCancel()` here that you then pass down to established sockets (creating a new derived context for all of them), that way live connections will be automatically canceled at shutdown or config changes

cre4ture replied:
thanks. done. I also added a test for this specific feature.

akernet commented on port-forwarder/fwd_tcp.go:
```diff
@@ -0,0 +1,230 @@
+		buf := make([]byte, 1*(1<<20))
+		for {
+			select {
+			case <-dataTransferCtx.Done():
```
I think you can handle this in a separate goroutine and call `from.Close()` and `to.Cancel()`

cre4ture replied:
Thanks for pointing out. I realized that it is actually even redundant. Thus I completely removed it.

akernet reviewed (commented):
Really nice! Cool that you got the buffer reuse to work

cre4ture commented on examples/config.yml:
```diff
@@ -206,6 +206,10 @@ relay:
+  # When tun is set to "user" then a userspace tun is used instead of an official kernel supported one.
+  # This way nebula can be used without root privileges and offers port forwarding from and into
+  # the nebula network. "tun.disabled" needs to be false for this use-case.
+  user: false
```
This was the quickest way to make it run. But its true, its not very intuitive. I will give it a try. Thanks

cre4ture replied:
done :-)

cre4ture commented on examples/config.yml:
```diff
@@ -368,3 +372,22 @@ firewall:
+  #  protocols: [tcp, udp]
+  inbound:
+  # format of forward_address: <host/ip>:<port>
+  #- port: 5599
```
Will it ever be usefull to provide an IP-address additionally to the port here? I'm not an expert, but from what I understood, the only relevant ip-address to specify here is the own nebula-ip. This was the reason to reduce this setting just to the port. I would suggest to call it `listen_port`. Maybe this way its best?

cre4ture replied:
changed to `listen_port`. thanks

akernet replied:
I'm not sure we want to support it here but I believe that you can have more than one IP assigned to a Nebula certificate/host so it kinda makes sense to leave it configurable.

cre4ture commented on cmd/nebula/main.go:
```diff
@@ -52,16 +59,61 @@ func main() {
-		os.Exit(1)
-	}
+	userspace_tun := c.GetBool("tun.user", false)
+	if userspace_tun {
```
I had a quick look. But this is not easily possible. We instanciate the `service`  class a few lines below with service.New(). Inside of this, the `nebula.Main()` is called. So we would need to refactor the instaciation of the `service`. I currently don't see the big value to justify this.

cre4ture commented on examples/config.yml:
```diff
@@ -368,3 +372,22 @@ firewall:
+port_forwarding:
+  outbound:
+  # format of local and remote address: <host/ip>:<port>
+  #- local_address: 127.0.0.1:3399
```
changed accordingly. thanks

cre4ture commented:
@akernet please re-review. and it seems that you need to also sign the cla as I cherry picked from your branch to honor your contribution. :-)

cre4ture committed:
support unsafe_routes for use with port_forwarding and user_tun

cre4ture committed:
performance impr.: avoid repeated allocation of "lastTick" on heap

cre4ture committed:
performance 3: avoid repeated allocation of []byte{1} on heap

cre4ture committed:
performance4: use buffer.view based channels instead of pipe

cre4ture committed:
Revert "Add script for speedtesting port forwarding"

This reverts commit c245a309666be91afc603cb6b3a8ad706df7fbfc.

cre4ture committed:
modified and extended speedtest

cre4ture committed:
add automated functional tests for port forwarding

cre4ture committed:
TCP/UDP port fwd. for disabled tun with config reload support

- reload config supported
  - killing existing live connections for removed fwd. configs

cre4ture committed:
fix fmt

cre4ture commented:
@johnmaguire Can you approve another CI run on this branch? It should be green accroding to the run in my fork: https://github.com/cre4ture/nebula/pull/1

cre4ture committed:
try to fix instability of the service level tests

cre4ture committed:
lets see if randomization of the port helps

cre4ture committed:
avoid panic due to writing to closed channel

cre4ture commented:
@johnmaguire thanks for running the CI. It has detected a instability that I do not fully understand. It seems that the tests get messed up due to using the same port  numers. But I explicitly made each test using its own port number. Do you know if the CI runns the same test multiple times in parallel? Otherwise I can't really explaint this.
For now I added some random number generator for additionally getting differences in the used port numbers. But its clear that this just reduces the chance for an issue and not completely solves it.

Anyway, could yoe please run the CI another time?
Thanks.

johnmaguire commented:
Hi @cre4ture - I don't believe that CI runs the same test in parallel multiple times. There is a test matrix which runs the tests on various OS. Looking at the failed CI, it appears it succeeded on windows-latest, macos-latest, linux with boringcrypto, but failed on ubuntu-linux. Oftentimes this is indicative of some kind of intermittent test failure (i.e. flaky test, maybe a race condition.)

> It seems that the tests get messed up due to using the same port numers.

I admittedly only took a very quick look, but I didn't find a log message that suggested a requested port was in use. Perhaps I missed it. Can you share what you're looking at?

cre4ture committed:
extend validity time range to avoid race conditions in CI

cre4ture commented:
> Hi @cre4ture - I don't believe that CI runs the same test in parallel multiple times. There is a test matrix which runs the tests on various OS in individual containers - so ports should not conflict. Looking at the failed CI, it appears it succeeded on windows-latest, macos-latest, linux with boringcrypto, but failed on ubuntu-linux. Oftentimes this is indicative of some kind of intermittent test failure (i.e. flaky test, maybe a race condition.)
> 
> > It seems that the tests get messed up due to using the same port numers.
> 
> I admittedly only took a very quick look, but I didn't find a log message that suggested a requested port was in use. Perhaps I missed it. Can you share what you're looking at?

@johnmaguire  I'm also not fully sure about this. An error message like you suggest would make it clear. But I can't explain why the failing test complains about not fitting ca-certificates otherwise.
What I'm looking at is this:
```
Invalid certificate from host" error="certificate validation failed: could not find ca for the certificate
```

and also this:

```
"Invalid certificate from host" error="certificate validation failed: certificate is expired"
```

But for the second one, I have an explanation meanwhile: the generated certificate for the test-services is only valid for 5 minutes. as the relevant test runns into a timeout after 10 minutes, its clear that this error occurs after 5 minutes.

Still, the first error I can't explain. The ca-key-pair is also generated freshly each time the test runs. From the ca-key-pair, the signed service key-pair and certificate is generated each time anew. So there should not be the possibility that these do not match. Expect, if one consideres that tests run in parallel an they accidentially connect with the wrong test-partner.

But I'm open for other ideas. Any help is welcome.

cre4ture committed:
ensure that node certs lifetime doesn't outlife ca cert lifetime

cre4ture committed:
consider injected name for certificate configuration

cre4ture committed:
try to unique node names in tests

cre4ture committed:
add name of the test to the certificate

cre4ture committed:
add logging prefix to differentiate the output in a test with 2 services

cre4ture committed:
tests: check for clean service shutdown and improve logging

cre4ture committed:
try to make it more stable by using channels and waitgroups

cre4ture committed:
improve stopping logic for UserDevice

cre4ture committed:
improving test code to get stability improved - still fails with stress

johnmaguire commented:
FYI we recently merged #1181 which caused conflicts. One conflict is caused by returning `*gonet.TCPConn` in lieu of `net.Conn`. Is it necessary to return the former, or we can continue to return the interface? Thanks!

cre4ture committed:
Merge remote-tracking branch 'origin/master' into feature/try_with_gvisor_stack

# Conflicts:
#	examples/go_service/main.go
#	service/service.go

cre4ture commented:
> FYI we recently merged #1181 which caused conflicts. One conflict is caused by returning `*gonet.TCPConn` in lieu of `net.Conn`. Is it necessary to return the former, or we can continue to return the interface? Thanks!

I merged the changes from main. It seems that the interface is OK.
Please tell me if you prefer a rebase, and/or a squash commit/merge.

cre4ture committed:
fix issue with survival of nebula service from previous testrun

cre4ture commented on udp/udp_linux.go:
```diff
@@ -315,6 +321,10 @@ func (u *StdConn) getMemInfo(meminfo *[unix.SK_MEMINFO_VARS]uint32) error {
 
 func (u *StdConn) Close() error {
 	//TODO: this will not interrupt the read loop
```
@johnmaguire I think I have the test stable now. At least on linux. The issue was a unclean shutdown of the nebula service at the end of each test which caused confusion in the following tests.

The change here in this file solved the issue.

cre4ture committed:
require instead of assert; fix missing close connectons in one test

cre4ture committed:
service: fix missing destruction of ipstack

cre4ture commented on service/service.go:
```diff
@@ -237,15 +243,49 @@ func (s *Service) Listen(network, address string) (net.Listener, error) {
-	return s.eg.Wait()
+	err := s.eg.Wait()
+
+	s.ipstack.Destroy()
```
Adding this line made the windows test stable. They no run 200+ in a row without issue.

cre4ture committed:
add comment to "unsafe_routes" initialisation

cre4ture committed:
add comment to performance improvement in user-tun

cre4ture committed:
add missing error handling when calling fwd factorie functions

cre4ture committed:
remove all sleeps in tests - no need for them

cre4ture committed:
nitpick: use atomic bool instead of bool for usynced thread access

cre4ture committed:
fix race condition where "CloseAndZero" is executed while still used

cre4ture commented on udp/udp_rio_windows.go:
```diff
@@ -323,6 +329,14 @@ func (u *RIOConn) Close() error {
 	windows.PostQueuedCompletionStatus(u.rx.iocp, 0, 0, nil)
 	windows.PostQueuedCompletionStatus(u.tx.iocp, 0, 0, nil)
 
+	u.rx.mu.Lock()                   // for waiting till active reader is done
```
Had to run the test 1000 times to achieve stable reproduction on my windows 11. Is fixed with this change.

cre4ture committed:
fix the closing of the linux udp reading loop using Shutdown

cre4ture commented:
@johnmaguire after extensive testing and a few findings I dare to claim that the tests are now stable on windows and linux. Can you please approve another workflow run on this PR? And have a look about what is still do be done to merge this?

ExplodingDragon commented:
@cre4ture Any new changes? Currently it seems to be incompatible with the master branch.

cre4ture committed:
Merge remote-tracking branch 'origin/master' into feature/try_with_gvisor_stack

# Conflicts:
#	service/service_test.go

cre4ture commented:
@ExplodingDragon  thanks for the notification. I merged the changes in.

maggie44 commented:
Is it possible to also add the ability to map two cidrs against each other? 

   - local_cidr: 10.0.0.0/24
      remote_cidr: 192.168.100.0/24

cre4ture commented:
> Is it possible to also add the ability to map two cidrs against each other?
> 
>     * local_cidr: 10.0.0.0/24
>       remote_cidr: 192.168.100.0/24

Hello maggie44,

thanks for your interesting question. I can't really answer that without further studying the documentation myself.
If nebula supports it, then its probably only possible with a real tun and thus root access.
I think this would be a good starting point for you:
https://nebula.defined.net/docs/config/tun/#tunroutes

hope this helps you a step forward.

maggie44 commented:
I meant in relation to this PR. Based on the example in the initial post it is configured like this:

```
port_forwarding:
  outbound:
  # format of local and remote address: <host/ip>:<port>
  #- local_address: 127.0.0.1:3399
  #  remote_address: 192.168.100.92:4499
     # format of protocols lists (yml-list): [tcp], [udp], [tcp, udp]
  #  protocols: [tcp, udp]
```

I was querying whether it could also be configured like this:

```
port_forwarding:
  outbound:
  #- local_address: 10.0.0.0/24:3399
  #  remote_address: 192.168.100.0/24:4499
     # format of protocols lists (yml-list): [tcp], [udp], [tcp, udp]
  #  protocols: [tcp, udp]
```

But I see now that the local address would need an interface? Which would defeat the object of removing the TUN device.

maggie44 commented:
Best I can think of right now would be to have TCP forwarding from a local IP, like `127.0.0.1:3399`, routed through to an IP based on path. This is very much an API related feature. For example:

```
http://127.0.0.1:3399/192.168.100.0
```

Where `192.168.100.0` is a path. Traffic routes in to `127.0.0.1:3399` which avoids requiring the TUN, then base on path `192.168.100.0` proxies traffic to the required host. That would allow opening TCP connections to an unlimited amount of connected hosts (I'm intentionally avoiding talking about UDP in relation to this, as the paths wouldn't be compatible). 

Straying off the original intent of this PR though, initially I was thinking of it only as a CIDR mapping which would have been a smaller change.

cre4ture commented:
@maggie44 Aha. I see the purpose of your question more clear now.
Actually I recalled that in the Feature request ( #1014) something similar like this is already mentioned which links to the feature request of having a SOCKS-Proxy (#915).
By following this approach, I thing you can achieve what you like. Just install a SOCKS proxy or similar on the remote host.
This way, one can probably also achieve the port-range "forwarding" request of @saket424 in the feature request.

maggie44 commented:
Thanks @cre4ture.

Here is the sort of things I had in mind. Looks like it's possible with the gVisor `Service` already in the main branch:

```
func startHTTPReverseProxy(ctx context.Context, svc *service.Service) error {
    mux := http.NewServeMux()
    server := &http.Server{
        Addr:    ":3399",
        Handler: mux,
    }

    // Create a custom transport with the custom dial context
    transport := &http.Transport{
        DialContext: func(ctx context.Context, network, addr string) (net.Conn, error) {
            return svc.DialContext(ctx, network, addr)
        },
    }

    mux.HandleFunc("/proxy/", func(w http.ResponseWriter, r *http.Request) {
        // Extract the target from query parameters
        target := r.URL.Query().Get("target")
        if target == "" {
            http.Error(w, "missing target parameter", http.StatusBadRequest)
            return
        }

        // Parse the target URL
        targetURL, err := url.Parse(target)
        if err != nil {
            http.Error(w, fmt.Sprintf("invalid target parameter: %s", err), http.StatusBadRequest)
            return
        }

        // Create a reverse proxy
        proxy := &httputil.ReverseProxy{
            Rewrite: func(r *httputil.ProxyRequest) {
                // Use only the target's path and query parameters. SetURL sets the majority of the URL fields.
                // Path and RawQuery replace the path to override the default which is to append the request
                // path to the target.
                r.SetURL(targetURL)
                r.Out.URL.Path = targetURL.Path
                r.Out.URL.RawQuery = targetURL.RawQuery
            },
            Transport: transport,
        }

        proxy.ServeHTTP(w, r)
    })

    go func() {
        <-ctx.Done()
        log.Println("shutting down server")
        _ = server.Shutdown(context.Background())
    }()

    log.Printf("starting server on %s", server.Addr)
    return server.ListenAndServe()
}
```

niowniow commented:
First of all, thank you for this contribution. It works very well for me.

I just wanted to note that port forwarding from the same node to itself does not work with this solution. I know that it's unnecessary because I can directly connect to the port on the same node, but by accident it happened to me. The following does not work on my machine with nebula/dial_address 192.168.100.3.

```
port_forwarding:
  enable_without_rules: false
  outbound:
  - listen_address: "127.0.0.1:5558"
    dial_address: "192.168.100.3:8880"
    protocols: [tcp, udp]
  inbound:
  - listen_port: 8880
    dial_address: 0.0.0.0:8880
    protocols: [tcp, udp]
```

I don't know enough details of how it works, so I can't debug it. It tries to connect but gets stuck.
```
wget http://localhost:5558/
--2024-12-19 10:05:30--  http://localhost:5558/
Resolving localhost (localhost)... 127.0.0.1
Connecting to localhost (localhost)|127.0.0.1|:5558... connected.
HTTP request sent, awaiting response...
```

Feel free to ignore, since I can bypass the issue by connecting directly to the port on the same machine.

cre4ture commented:
> dial_address

Hi @niowniow,

thanks for testing the implementation on this PR and your feedback.
I assume that "192.168.100.3" is your own IP in the nebula network. Right?

So, I can't really tell if this is supported or not as it depends on a part of the code that this PR doesn't touch. I think.
But did you check if dialing to "127.0.0.1" instead of "0.0.0.0" does help?
I currently getting the impression that there is the mixup of listen and dial IP.

cre4ture committed:
Merge remote-tracking branch 'origin/master' into feature/try_with_gvisor_stack

# Conflicts:
#	overlay/user.go
#	service/service_test.go
#	udp/udp_linux.go

cre4ture commented:
@johnmaguire @akernet 
Hello guys, is there any idea from your side when this PR can be reviewed from your side?

johnmaguire commented:
Hi @cre4ture, there's been a big ongoing push on #6 that has prevented us from looking too closely at this PR. I'm optimistic that we can start working through a backlog of PRs that deserve reviews soon.

ExplodingDragon commented:
@cre4ture Any new changes? Currently it seems to be incompatible with the master branch.

There appears to be a potential memory leak. (Version: 1.9.5 , Patch:  03630fbf3517a636777de0e96cfe2e2c1ea58086)

After restarting Nebula, the memory usage has changed.

![](https://github.com/user-attachments/assets/0c003f60-8c3a-4cfc-9c0f-bd49168b09d7)

cre4ture committed:
Merge remote-tracking branch 'origin/master' into feature/try_with_gvisor_stack

cre4ture committed:
fix golangci-lint issues

cre4ture committed:
fix leaking go-routine

cre4ture committed:
Merge remote-tracking branch 'origin/master' into feature/try_with_gvisor_stack

ExplodingDragon commented:
@cre4ture  all right, i think it's fixed. thanks.
<img width="1016" height="214" alt="" src="https://github.com/user-attachments/assets/8a3fa998-4bdf-4334-b5d1-14b0db6e79ef" />

cre4ture commented:
> should this be included in the main `nebula` binary or released under a different binary, say `nebula-proxy`/`nebula-pf`/etc.

ExplodingDragon commented:
@nbrownus Any updates on this? I've been using this PR on 20+ devices for a while now, and everything looks good. Would be great to get this merged upstream.
